File: Sources/DupliDetect/Fingerprint.swift
import Foundation
import Accelerate

/// A compact acoustic summary of a file: one 32-bit sub-fingerprint per frame.
///
/// The scheme is the Haitsma-Kalker robust hash. For each frame the spectrum is
/// folded into 33 logarithmically spaced bands, and each bit records whether the
/// energy difference between neighbouring bands rose or fell relative to the
/// previous frame. Because every bit comes from a *difference of differences*,
/// the result is invariant to overall volume and largely unaffected by the
/// artefacts of lossy encoding — two encodes of the same master produce nearly
/// the same bits even at very different bitrates.
public struct Fingerprint: Hashable {
    public let values: [UInt32]

    /// The average spectral shape of the file: one quantised value per band,
    /// measured relative to the loudest band so it is volume-invariant.
    /// Describes *what* the audio sounds like rather than how it changes, which
    /// is the only thing left to compare when the audio never changes.
    public let shapeProfile: [UInt8]

    /// Median frame-to-frame change in log band energy. Near zero for a held
    /// tone or a drone; comfortably above it for music and speech.
    public let flux: Double

    /// True when the audio barely changes from frame to frame, which makes the
    /// difference-based bits in `values` pure rounding noise.
    public var isStationary: Bool { flux < FingerprintEngine.stationaryFlux }
    /// Seconds of audio each sub-fingerprint advances.
    public static let secondsPerFrame = Double(FingerprintEngine.hopSize) / AudioDecoder.fingerprintRate

    public var duration: Double { Double(values.count) * Fingerprint.secondsPerFrame }
    public var isUsable: Bool { values.count >= FingerprintEngine.minimumFrames }
}

/// A fingerprint plus what the engine learned about the audio while making it.
public struct AudioAnalysis {
    public let fingerprint: Fingerprint
    /// Seconds of near-silence trimmed from the head and tail before fingerprinting.
    public let leadingSilence: Double
    public let trailingSilence: Double
}

public final class FingerprintEngine {

    static let frameSize = 4096
    static let hopSize = 512
    static let bandCount = 33          // 33 bands -> 32 bits per frame
    static let lowFrequency = 300.0
    static let highFrequency = 3000.0
    /// Roughly 1.5 seconds of audio; below this there is nothing to compare.
    static let minimumFrames = 64

    /// Below this median spectral flux the audio is treated as stationary.
    /// Measured with a wide margin either side: a held tone reads about 0.011
    /// as PCM and 0.044 once lossily encoded, while music and speech sit at
    /// 0.23 and above.
    static let stationaryFlux = 0.10

    private let log2n: vDSP_Length
    private let setup: FFTSetup
    private let window: [Float]
    private let bandEdges: [Int]

    private var realPart: [Float]
    private var imagPart: [Float]
    private var magnitudes: [Float]

    public init() {
        let n = FingerprintEngine.frameSize
        log2n = vDSP_Length(log2(Double(n)).rounded())
        setup = vDSP_create_fftsetup(log2n, FFTRadix(kFFTRadix2))!

        var hann = [Float](repeating: 0, count: n)
        vDSP_hann_window(&hann, vDSP_Length(n), Int32(vDSP_HANN_NORM))
        window = hann

        bandEdges = FingerprintEngine.makeBandEdges(frameSize: n,
                                                    sampleRate: AudioDecoder.fingerprintRate)

        realPart = [Float](repeating: 0, count: n / 2)
        imagPart = [Float](repeating: 0, count: n / 2)
        magnitudes = [Float](repeating: 0, count: n / 2)
    }

    deinit { vDSP_destroy_fftsetup(setup) }

    /// Logarithmically spaced FFT bin boundaries covering 300 Hz - 3 kHz.
    private static func makeBandEdges(frameSize: Int, sampleRate: Double) -> [Int] {
        let binsPerHz = Double(frameSize) / sampleRate
        var edges: [Int] = []
        edges.reserveCapacity(bandCount + 1)
        let ratio = log(highFrequency / lowFrequency)
        for i in 0...bandCount {
            let frequency = lowFrequency * exp(ratio * Double(i) / Double(bandCount))
            var bin = Int((frequency * binsPerHz).rounded())
            bin = min(max(bin, 1), frameSize / 2 - 1)
            // Guarantee every band owns at least one bin.
            if let last = edges.last, bin <= last { bin = last + 1 }
            edges.append(min(bin, frameSize / 2 - 1))
        }
        return edges
    }

    /// Computes the fingerprint of already-decoded mono audio, and reports the
    /// silence trimmed off each end so callers can tell how much real audio
    /// a file actually holds.
    public func analyze(samples: [Float], sampleRate: Double = AudioDecoder.fingerprintRate) -> AudioAnalysis {
        let range = FingerprintEngine.contentRange(samples)
        let leading = Double(range.lowerBound) / sampleRate
        let trailing = Double(samples.count - range.upperBound) / sampleRate
        return AudioAnalysis(fingerprint: fingerprint(samples: samples),
                             leadingSilence: leading,
                             trailingSilence: trailing)
    }

    /// Computes the fingerprint of already-decoded mono audio.
    public func fingerprint(samples: [Float]) -> Fingerprint {
        let trimmed = FingerprintEngine.trimSilence(samples)
        let n = FingerprintEngine.frameSize
        let empty = Fingerprint(values: [], shapeProfile: [], flux: .greatestFiniteMagnitude)
        guard trimmed.count >= n + FingerprintEngine.hopSize else { return empty }

        let frameCount = (trimmed.count - n) / FingerprintEngine.hopSize + 1
        var bandEnergies = [Float](repeating: 0, count: FingerprintEngine.bandCount)
        var previousEnergies: [Float]? = nil
        var values: [UInt32] = []
        var frameFlux: [Float] = []
        var bandTotals = [Double](repeating: 0, count: FingerprintEngine.bandCount)
        var shapeFrames = 0
        values.reserveCapacity(frameCount)

        var windowed = [Float](repeating: 0, count: n)

        for frame in 0..<frameCount {
            let start = frame * FingerprintEngine.hopSize
            trimmed.withUnsafeBufferPointer { source in
                vDSP_vmul(source.baseAddress! + start, 1, window, 1, &windowed, 1, vDSP_Length(n))
            }
            spectrum(of: windowed, into: &bandEnergies)

            // Spectral shape: which of two neighbouring bands is louder, within
            // this one frame. Scaling every band by the same factor cannot
            // change the answer, so this is volume-invariant like `values` —
            // but it survives audio that never changes.
            for m in 0..<FingerprintEngine.bandCount {
                bandTotals[m] += Double(bandEnergies[m])
            }
            shapeFrames += 1

            if let previous = previousEnergies {
                var bits: UInt32 = 0
                var change: Float = 0
                for m in 0..<32 {
                    let current = bandEnergies[m] - bandEnergies[m + 1]
                    let earlier = previous[m] - previous[m + 1]
                    if current - earlier > 0 { bits |= (1 << UInt32(m)) }
                }
                for m in 0..<FingerprintEngine.bandCount {
                    change += abs(bandEnergies[m] - previous[m])
                }
                values.append(bits)
                frameFlux.append(change / Float(FingerprintEngine.bandCount))
            }
            previousEnergies = bandEnergies
        }

        return Fingerprint(values: values,
                           shapeProfile: FingerprintEngine.shape(totals: bandTotals, frames: shapeFrames),
                           flux: Double(FingerprintEngine.median(frameFlux)))
    }

    /// Builds the spectral-shape template.
    ///
    /// Each band's mean log energy is expressed relative to the loudest band,
    /// which removes overall volume, then clamped to an 80 dB window and
    /// quantised. Clamping matters as much as the rest: bands holding nothing
    /// but noise floor land on the same floor value in every file, so the
    /// differences between one codec's idea of silence and another's stop
    /// counting as differences in the audio.
    static func shape(totals: [Double], frames: Int) -> [UInt8] {
        guard frames > 0, !totals.isEmpty else { return [] }

        let means = totals.map { $0 / Double(frames) }
        guard let peak = means.max() else { return [] }

        // Band energies are log10 of power, so 8.0 spans 80 dB.
        let window = 8.0
        return means.map { mean in
            let relative = min(0, max(-window, mean - peak))
            return UInt8(((relative + window) / window * 255).rounded())
        }
    }

    /// Distance between two shape templates, 0 (identical) to 1.
    /// Only meaningful when both files are stationary.
    ///
    /// Each band is weighted by how much energy either file puts there, so the
    /// comparison is driven by the bands that actually carry the sound. Without
    /// that weighting, two quite different tones look similar merely because
    /// most of the spectrum is empty in both of them.
    public static func shapeDistance(_ a: [UInt8], _ b: [UInt8]) -> Double {
        guard a.count == b.count, !a.isEmpty else { return 1 }

        var weightedDifference = 0.0
        var totalWeight = 0.0
        for i in 0..<a.count {
            let valueA = Double(a[i]), valueB = Double(b[i])
            let weight = max(valueA, valueB) / 255.0
            weightedDifference += weight * abs(valueA - valueB)
            totalWeight += weight
        }
        guard totalWeight > 0 else { return 0 }
        return weightedDifference / totalWeight / 255.0
    }

    static func median(_ values: [Float]) -> Float {
        guard !values.isEmpty else { return .greatestFiniteMagnitude }
        let sorted = values.sorted()
        return sorted[sorted.count / 2]
    }

    /// Real FFT of one windowed frame, accumulated into log band energies.
    private func spectrum(of windowed: [Float], into bands: inout [Float]) {
        let n = FingerprintEngine.frameSize
        let half = n / 2

        realPart.withUnsafeMutableBufferPointer { realBuffer in
            imagPart.withUnsafeMutableBufferPointer { imagBuffer in
                var split = DSPSplitComplex(realp: realBuffer.baseAddress!,
                                            imagp: imagBuffer.baseAddress!)
                windowed.withUnsafeBufferPointer { source in
                    source.baseAddress!.withMemoryRebound(to: DSPComplex.self, capacity: half) { complex in
                        vDSP_ctoz(complex, 2, &split, 1, vDSP_Length(half))
                    }
                }
                vDSP_fft_zrip(setup, &split, 1, log2n, FFTDirection(FFT_FORWARD))
                // Power spectrum; the 1/4 factor undoes vDSP's zrip scaling.
                vDSP_zvmags(&split, 1, &magnitudes, 1, vDSP_Length(half))
            }
        }

        var scale: Float = 0.25
        vDSP_vsmul(magnitudes, 1, &scale, &magnitudes, 1, vDSP_Length(half))

        for band in 0..<FingerprintEngine.bandCount {
            let lower = bandEdges[band]
            let upper = max(bandEdges[band + 1], lower + 1)
            var sum: Float = 0
            magnitudes.withUnsafeBufferPointer { buffer in
                vDSP_sve(buffer.baseAddress! + lower, 1, &sum, vDSP_Length(upper - lower))
            }
            // Log energy keeps quiet passages as informative as loud ones.
            bands[band] = log10f(sum + 1e-9)
        }
    }

    /// Drops near-silent head and tail so padding differences do not shift alignment.
    static func trimSilence(_ samples: [Float]) -> ArraySlice<Float> {
        let range = contentRange(samples)
        return samples[range]
    }

    /// The span of `samples` that actually carries audio.
    static func contentRange(_ samples: [Float]) -> Range<Int> {
        guard !samples.isEmpty else { return 0..<0 }

        var peak: Float = 0
        vDSP_maxmgv(samples, 1, &peak, vDSP_Length(samples.count))
        guard peak > 1e-5 else { return 0..<samples.count }
        let threshold = peak * 0.005

        var start = 0
        while start < samples.count && abs(samples[start]) < threshold { start += 1 }
        var end = samples.count - 1
        while end > start && abs(samples[end]) < threshold { end -= 1 }
        guard end > start else { return 0..<samples.count }
        return start..<(end + 1)
    }
}

File: rust/dd-core/src/fingerprint.rs
//! Acoustic fingerprinting, ported from Sources/DupliDetect/Fingerprint.swift.
//!
//! Every constant here is chosen to reproduce the macOS implementation exactly,
//! not because it reads well. Three in particular are load-bearing and are
//! explained where they appear: the Hann window's normalisation, the absence of
//! a spectrum scale factor, and the fact that FFT bin 0 is never read.

use realfft::{RealFftPlanner, RealToComplex};
use std::sync::Arc;

/// Rate the fingerprinter works at.
pub const FINGERPRINT_RATE: f64 = 11025.0;

pub const FRAME_SIZE: usize = 4096;
pub const HOP_SIZE: usize = 512;
pub const BAND_COUNT: usize = 33; // 33 bands -> 32 bits per frame
pub const LOW_FREQUENCY: f64 = 300.0;
pub const HIGH_FREQUENCY: f64 = 3000.0;
/// Roughly 1.5 seconds of audio; below this there is nothing to compare.
pub const MINIMUM_FRAMES: usize = 64;
/// Below this median spectral flux the audio is treated as stationary.
pub const STATIONARY_FLUX: f64 = 0.10;

/// Seconds of audio each sub-fingerprint advances.
pub fn seconds_per_frame() -> f64 {
    HOP_SIZE as f64 / FINGERPRINT_RATE
}

#[derive(Clone, Debug, PartialEq)]
pub struct Fingerprint {
    pub values: Vec<u32>,
    pub shape_profile: Vec<u8>,
    pub flux: f64,
}

impl Fingerprint {
    pub fn empty() -> Self {
        Fingerprint { values: Vec::new(), shape_profile: Vec::new(), flux: f64::MAX }
    }
    pub fn duration(&self) -> f64 {
        self.values.len() as f64 * seconds_per_frame()
    }
    pub fn is_usable(&self) -> bool {
        self.values.len() >= MINIMUM_FRAMES
    }
    pub fn is_stationary(&self) -> bool {
        self.flux < STATIONARY_FLUX
    }
}

#[derive(Clone, Debug)]
pub struct AudioAnalysis {
    pub fingerprint: Fingerprint,
    pub leading_silence: f64,
    pub trailing_silence: f64,
}

pub struct FingerprintEngine {
    fft: Arc<dyn RealToComplex<f32>>,
    window: Vec<f32>,
    band_edges: Vec<usize>,
    scratch: Vec<realfft::num_complex::Complex<f32>>,
    spectrum_out: Vec<realfft::num_complex::Complex<f32>>,
    magnitudes: Vec<f32>,
}

impl Default for FingerprintEngine {
    fn default() -> Self {
        Self::new()
    }
}

impl FingerprintEngine {
    pub fn new() -> Self {
        let mut planner = RealFftPlanner::<f32>::new();
        let fft = planner.plan_fft_forward(FRAME_SIZE);
        let scratch = fft.make_scratch_vec();
        let spectrum_out = fft.make_output_vec();
        FingerprintEngine {
            fft,
            window: make_window(),
            band_edges: make_band_edges(FRAME_SIZE, FINGERPRINT_RATE),
            scratch,
            spectrum_out,
            magnitudes: vec![0.0; FRAME_SIZE / 2],
        }
    }

    pub fn window(&self) -> &[f32] {
        &self.window
    }
    pub fn band_edges(&self) -> &[usize] {
        &self.band_edges
    }

    /// Fingerprint plus the silence trimmed from each end.
    pub fn analyze(&mut self, samples: &[f32], sample_rate: f64) -> AudioAnalysis {
        let range = content_range(samples);
        let leading = range.0 as f64 / sample_rate;
        let trailing = (samples.len() - range.1) as f64 / sample_rate;
        AudioAnalysis {
            fingerprint: self.fingerprint(samples),
            leading_silence: leading,
            trailing_silence: trailing,
        }
    }

    pub fn fingerprint(&mut self, samples: &[f32]) -> Fingerprint {
        let (start, end) = content_range(samples);
        let trimmed = &samples[start..end];
        if trimmed.len() < FRAME_SIZE + HOP_SIZE {
            return Fingerprint::empty();
        }

        let frame_count = (trimmed.len() - FRAME_SIZE) / HOP_SIZE + 1;
        let mut band_energies = [0.0f32; BAND_COUNT];
        let mut previous: Option<[f32; BAND_COUNT]> = None;
        let mut values: Vec<u32> = Vec::with_capacity(frame_count);
        let mut frame_flux: Vec<f32> = Vec::with_capacity(frame_count);
        let mut band_totals = [0.0f64; BAND_COUNT];
        let mut shape_frames = 0usize;
        let mut windowed = vec![0.0f32; FRAME_SIZE];

        for frame in 0..frame_count {
            let offset = frame * HOP_SIZE;
            for i in 0..FRAME_SIZE {
                windowed[i] = trimmed[offset + i] * self.window[i];
            }
            self.spectrum(&windowed, &mut band_energies);

            for m in 0..BAND_COUNT {
                band_totals[m] += band_energies[m] as f64;
            }
            shape_frames += 1;

            if let Some(prev) = previous {
                let mut bits: u32 = 0;
                let mut change: f32 = 0.0;
                for m in 0..32 {
                    let current = band_energies[m] - band_energies[m + 1];
                    let earlier = prev[m] - prev[m + 1];
                    if current - earlier > 0.0 {
                        bits |= 1 << (m as u32);
                    }
                }
                for m in 0..BAND_COUNT {
                    change += (band_energies[m] - prev[m]).abs();
                }
                values.push(bits);
                frame_flux.push(change / BAND_COUNT as f32);
            }
            previous = Some(band_energies);
        }

        Fingerprint {
            values,
            shape_profile: shape(&band_totals, shape_frames),
            flux: median(&frame_flux) as f64,
        }
    }

    /// Real FFT of one windowed frame, folded into log band energies.
    pub fn spectrum(&mut self, windowed: &[f32], bands: &mut [f32; BAND_COUNT]) {
        let mut input = windowed.to_vec();
        self.fft
            .process_with_scratch(&mut input, &mut self.spectrum_out, &mut self.scratch)
            .expect("fft");

        // Swift applies 0.25 here because vDSP_fft_zrip returns 2x the
        // mathematical DFT, so squaring gives 4x. realfft already produces the
        // unscaled mathematical DFT, so squaring is already correct and the
        // 0.25 must NOT be carried across.
        for k in 0..FRAME_SIZE / 2 {
            let c = self.spectrum_out[k];
            self.magnitudes[k] = c.re * c.re + c.im * c.im;
        }
        // vDSP packs DC and Nyquist together into bin 0, making magnitudes[0]
        // meaningless there. Harmless in both implementations: the lowest band
        // edge is clamped to 1 and 300 Hz lands at bin 111, so bin 0 is never
        // summed. Do not "fix" this.

        for band in 0..BAND_COUNT {
            let lower = self.band_edges[band];
            let upper = self.band_edges[band + 1].max(lower + 1);
            let mut sum = 0.0f32;
            for k in lower..upper {
                sum += self.magnitudes[k];
            }
            // Log energy keeps quiet passages as informative as loud ones.
            // The epsilon means this is NOT scale-invariant near the noise
            // floor, which is why the spectrum scale above has to be right.
            bands[band] = (sum + 1e-9).log10();
        }
    }
}

/// The Hann window vDSP produces for `vDSP_HANN_NORM`.
///
/// From the SDK header: `W = .816496580927726` (sqrt(2/3)) and
/// `C[n] = W * (1 - cos(2*pi*n/N))` — periodic (`/N`, not `/(N-1)`), giving a
/// peak of 1.633 rather than 1.0. A textbook `0.5 * (1 - cos(...))` would be
/// 0.612x in amplitude and 0.375x in power, which shifts every band energy.
///
/// One deliberate divergence: vDSP evaluates `1 - cos(x)` in f32, which loses
/// most of its significant digits to cancellation for small `x` — at n = 1 it
/// yields 1.192e-6 where the true value is 1.177e-6, a 1.3% error. This uses
/// the algebraically identical `2 sin²(x/2)`, which has no cancellation at all.
///
/// That is not merely more accurate, it is the only portable choice. Matching
/// vDSP bit for bit would mean reproducing its f32 cancellation, which
/// amplifies last-ulp differences between platform `cosf` implementations by
/// four orders of magnitude — Windows and Linux could then disagree with each
/// other about the window, and therefore about fingerprints. This form gives
/// every platform the same answer.
///
/// The divergence is immaterial: peak amplitude is bit-identical to vDSP, the
/// largest absolute difference anywhere is 2.4e-7 against a peak of 1.633
/// (1.5e-7 relative), and the only coefficients differing by more than 0.1%
/// are twelve whose values are below 7.8e-5 and so contribute nothing. Tier 0
/// asserts the band energies that result, and Tier 1 asserts the fingerprints.
pub fn make_window() -> Vec<f32> {
    const W: f64 = 0.816_496_580_927_726;
    let n = FRAME_SIZE as f64;
    (0..FRAME_SIZE)
        .map(|i| {
            let half = std::f64::consts::PI * i as f64 / n; // x/2 where x = 2*pi*i/N
            let sin_half = half.sin();
            (W * 2.0 * sin_half * sin_half) as f32
        })
        .collect()
}

/// Logarithmically spaced FFT bin boundaries covering 300 Hz - 3 kHz.
pub fn make_band_edges(frame_size: usize, sample_rate: f64) -> Vec<usize> {
    let bins_per_hz = frame_size as f64 / sample_rate;
    let ratio = (HIGH_FREQUENCY / LOW_FREQUENCY).ln();
    let mut edges: Vec<usize> = Vec::with_capacity(BAND_COUNT + 1);
    for i in 0..=BAND_COUNT {
        let frequency = LOW_FREQUENCY * (ratio * i as f64 / BAND_COUNT as f64).exp();
        // Swift's .rounded() is half-away-from-zero, matching f64::round().
        let mut bin = (frequency * bins_per_hz).round() as usize;
        bin = bin.clamp(1, frame_size / 2 - 1);
        if let Some(&last) = edges.last() {
            if bin <= last {
                bin = last + 1;
            }
        }
        edges.push(bin.min(frame_size / 2 - 1));
    }
    edges
}

/// The span of `samples` that actually carries audio.
pub fn content_range(samples: &[f32]) -> (usize, usize) {
    if samples.is_empty() {
        return (0, 0);
    }
    let peak = samples.iter().fold(0.0f32, |acc, v| acc.max(v.abs()));
    if peak <= 1e-5 {
        return (0, samples.len());
    }
    let threshold = peak * 0.005;

    let mut start = 0;
    while start < samples.len() && samples[start].abs() < threshold {
        start += 1;
    }
    let mut end = samples.len() - 1;
    while end > start && samples[end].abs() < threshold {
        end -= 1;
    }
    if end <= start {
        return (0, samples.len());
    }
    (start, end + 1)
}

/// Spectral-shape template: mean band energy relative to the loudest band,
/// clamped to an 80 dB window and quantised.
pub fn shape(totals: &[f64], frames: usize) -> Vec<u8> {
    if frames == 0 || totals.is_empty() {
        return Vec::new();
    }
    let means: Vec<f64> = totals.iter().map(|t| t / frames as f64).collect();
    let peak = means.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let window = 8.0;
    means
        .iter()
        .map(|mean| {
            let relative = (mean - peak).max(-window).min(0.0);
            (((relative + window) / window * 255.0).round()) as u8
        })
        .collect()
}

/// Weighted distance between two shape templates, 0 (identical) to 1.
pub fn shape_distance(a: &[u8], b: &[u8]) -> f64 {
    if a.len() != b.len() || a.is_empty() {
        return 1.0;
    }
    let mut weighted_difference = 0.0;
    let mut total_weight = 0.0;
    for i in 0..a.len() {
        let (va, vb) = (a[i] as f64, b[i] as f64);
        let weight = va.max(vb) / 255.0;
        weighted_difference += weight * (va - vb).abs();
        total_weight += weight;
    }
    if total_weight <= 0.0 {
        return 0.0;
    }
    weighted_difference / total_weight / 255.0
}

/// Upper median, matching Swift's `sorted[count / 2]`.
pub fn median(values: &[f32]) -> f32 {
    if values.is_empty() {
        return f32::MAX;
    }
    let mut sorted = values.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    sorted[sorted.len() / 2]
}
